File: demo-app/src/main/java/cn/dozyx/template/activity/LaunchModeTest.java
package cn.dozyx.template.activity;

import android.app.ActivityManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import cn.dozyx.template.base.BaseShowResultActivity;

/**
 * 通过 adb shell dumpsys activity activities 命令查看 Activity 栈信息
 */
public class LaunchModeTest extends BaseShowResultActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setText(this.getClass().getSimpleName());
        ActivityManager activityManager =
                (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);
    }

    @Override
    protected String[] getButtonText() {
        return new String[]{
                "standard", "singleTop", "singleTask", "singleInstance", "singleTask with taskAffinity"
        };
    }

    @Override
    public void onButton1() {
        callActivity(StandardLaunchModeTest.class);
    }

    @Override
    public void onButton2() {
        callActivity(SingleTopLaunchModeTest.class);
    }

    @Override
    public void onButton3() {
        callActivity(SingleTaskLaunchModeTest.class);
    }

    @Override
    public void onButton4() {
        callActivity(SingleInstanceLaunchModeTest.class);
    }

    @Override
    public void onButton5() {
        Intent intent = new Intent(this, StandardLaunchModeTest.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        finish();
        startActivity(intent);
    }

    private void callActivity(Class activity) {
        startActivity(new Intent(this, activity));
    }

    public static final class StandardLaunchModeTest extends LaunchModeTest {

    }

    public static final class SingleTopLaunchModeTest extends LaunchModeTest {

    }

    public static final class SingleTaskLaunchModeTest extends LaunchModeTest {

    }

    public static final class SingleInstanceLaunchModeTest extends LaunchModeTest {

    }

    public static final class SingleTaskLaunchModeTestWithTaskAffinity extends LaunchModeTest {

    }


}

File: demo-app/src/main/java/cn/dozyx/template/activity/LaunchModeTest.kt
package cn.dozyx.template.activity

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import cn.dozyx.template.base.Action

import cn.dozyx.template.base.BaseShowResultActivity
import cn.dozyx.template.base.BaseTestActivity
import timber.log.Timber

/**
 * 通过 adb shell dumpsys activity activities 命令查看 Activity 栈信息
 */
open class LaunchModeTest : BaseTestActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Timber.d("LaunchModeTest.onCreate ${this.javaClass.name} $taskId")
        addAction(object : Action("standard") {
            override fun run() {
                callActivity(StandardLaunchModeTest::class.java)
            }
        })
        addAction(object : Action("singleTop") {
            override fun run() {
                callActivity(SingleTopLaunchModeTest::class.java)
            }
        })
        addAction(object : Action("singleTask") {
            override fun run() {
                callActivity(SingleTaskLaunchModeTest::class.java)
            }
        })
        addAction(object : Action("singleInstance") {
            override fun run() {
                callActivity(SingleInstanceLaunchModeTest::class.java)
            }
        })
        addAction(object : Action("singleTask with taskAffinity") {
            override fun run() {
                val intent = Intent(this@LaunchModeTest, StandardLaunchModeTest::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
            }
        })

        addAction(object : Action("singleTask start standard") {
            override fun run() {
                val intent = Intent(this@LaunchModeTest, StandardLaunchModeTest::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
            }
        })

        addAction(object : Action("singleTask start singleTop") {
            override fun run() {
                val intent = Intent(this@LaunchModeTest, SingleTopLaunchModeTest::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
            }
        })

        addAction(object : Action("singleTask start singleTask") {
            override fun run() {
                val intent = Intent(this@LaunchModeTest, SingleTaskLaunchModeTest::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
            }
        })
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        Timber.d("LaunchModeTest.onNewIntent $taskId")
    }


    private fun callActivity(activity: Class<*>) {
        startActivity(Intent(this, activity))
    }

    override fun onDestroy() {
        Timber.d("LaunchModeTest.onDestroy ${this.javaClass.name} $taskId")
        super.onDestroy()
    }

    class StandardLaunchModeTest : LaunchModeTest()

    class SingleTopLaunchModeTest : LaunchModeTest()

    class SingleTaskLaunchModeTest : LaunchModeTest()

    class SingleInstanceLaunchModeTest : LaunchModeTest()

    class SingleTaskLaunchModeTestWithTaskAffinity : LaunchModeTest()


}
